Qiskit/qiskit-terra
# Add display of internal circuits for IfElseOps and WhileLoopOps to mpl circuit drawer

### Summary

This PR adds display of the circuits inside `IfElseOps` to the `mpl` circuit drawer

### Details and comments

UPDATED: Now boxes properly fold onto the next line and no longer a WIP.

UPDATE 2: Added WhileLoopOps

This PR is intended to be merged after #10096. It should accomplish the following

1. First, do no harm. All circuits without `IfElseOp`s should display EXACTLY as they do in main.
2. Whether created directly as `IfElseOp` or through `qc.if_test` all if-else should display with a box with the if and else circuits inside.

Testers Wanted!

Since the circuits inside the if-else can be very large and complex, it's difficult to test all the scenarios. If you have a qiskit-terra development environment on your local system, just create a new branch out of main, and copy `visualization/circuit/utils.py` and `visualization/circuit/matplotlib.py` into the branch. Things to test,
- Try lots of complex existing circuits to make sure 1. above is satisfied.
- Create lots of `if_test` examples with and without `else` and using different kinds and complexities of circuits.
- Set `fold=-1` to the draw command initially to test 2. This will do no folding.
- Set `fold` to a small number and see that the circuits properly wrap.

To do

Futures

- Add `switch`, `for`, and `while` control flow ops.
- Perhaps use some of this capability for showing sub-circuits or expanding gates built from circuits.
```
qr = QuantumRegister(4, "q")
cr = ClassicalRegister(3, "cr")
qc = QuantumCircuit(qr, cr)
qc.h(0)

with qc.if_test((cr[1], 1)) as _else:
    qc.x(0, label="X1111i").c_if(cr, 4)
    with qc.if_test((cr[2], 1)):
        qc.z(0)
        with qc.if_test((cr[1], 1)):
            qc.cx(0, 1)
            with qc.if_test((cr[2], 1)):
                qc.y(1)
                qc.z(2)
                with qc.if_test((cr[1], 1)):
                    qc.cx(0, 1)
with _else:
    qc.y(1)
    with qc.if_test((cr[2], 1)):
        qc.x(0)
        qc.x(1)
    qr1 = QuantumRegister(2, "qr1")
    cr1 = ClassicalRegister(2, "cr1")
    cr2 = ClassicalRegister(2, "cr2")
    circuit = QuantumCircuit(qr1, cr1, cr2)
    inst = QuantumCircuit(2, 2, name="Inst").to_instruction()
    qc.append(inst, [qr[0], qr[1]], [cr[0], cr[1]])

qc.x(0)
qc.draw(fold=30, style={"showindex": True})
```
![image](https://github.com/Qiskit/qiskit-terra/assets/16268251/72f7bad5-da19-491f-9521-46d391b48290)
Using fold=7, the above circuit looks like this,
![image](https://github.com/Qiskit/qiskit-terra/assets/16268251/e9974db3-32a6-4edb-8677-4d143a90b453)
Directly instantiating the `IfElseOp` instead of using `if_test` also works.
```
qr = QuantumRegister(4, "q")
cr = ClassicalRegister(3, "cr")
qc = QuantumCircuit(qr, cr)
qc.h(0)
qc.h(1)
qc.measure(0, 1)
qc.measure(1, 2)
qc.x(2)
qc.x(2, label="XLabel").c_if(cr, 2)

qr2 = QuantumRegister(3, "qr2")
qc2 = QuantumCircuit(qr2, cr)
qc2.x(1)
qc2.y(1)
qc2.z(0)
qc2.x(0, label="X1i").c_if(cr, 4)

qc.if_else((cr[1], 1), qc2, None, [0, 1, 2], [0, 1, 2])
qc.x(0, label="X1i")
qc.draw(style={"showindex": True}, cregbundle=True)
```
![image](https://github.com/Qiskit/qiskit-terra/assets/16268251/e2f0356d-ed9b-4dfc-b86b-a5c6bebf70d7)
